be/src/exec/scan_node.cpp

```diff
@@ -112,9 +112,11 @@ StatusOr<pipeline::MorselQueueFactoryPtr> ScanNode::convert_scan_range_to_morsel
         int scan_dop = std::min<int>(std::max<int>(1, morsel_queue->max_degree_of_parallelism()), pipeline_dop);
         int io_parallelism = scan_dop * io_tasks_per_scan_operator();
 
-        // If not so much morsels, try to assign morsel uniformly among operators to avoid data skew
-        if (!always_shared_scan() && scan_dop > 1 && dynamic_cast<pipeline::FixedMorselQueue*>(morsel_queue.get()) &&
-            morsel_queue->num_original_morsels() <= io_parallelism) {
+        // If session variable enable_shared_scan is false or not so much morsels,
+        // try to assign morsel uniformly among operators to avoid data skew
+        if (!enable_shared_scan || (allow_adaptive_disable_shared_scan() && scan_dop > 1 &&
```

Comment:
When the number of tablets is enough, we needn't assign morsel uniformly among operators.
```suggestion
        if (!enable_shared_scan && (allow_adaptive_disable_shared_scan() && scan_dop > 1 &&
```